### PR TITLE
Run commands in imageSpec

### PR DESCRIPTION
**Pull request by pingsutw**
# TL;DR
Add commands to ImageSpec, so people can download jars or any other packages. 

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Another example for the spark task
```python
import datetime
import random
from operator import add

import flytekit
from flytekit import Resources, task, workflow
from flytekitplugins.spark import Spark
from flytekit.image_spec.image_spec import ImageSpec

spark_image = ImageSpec(registry="pingsutw", apt_packages=["wget"])
spark_image.with_commands(["wget https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-aws/3.2.2/hadoop-aws-3.2.2.jar -P /opt/spark/jars",
                          "wget https://repo1.maven.org/maven2/com/amazonaws/aws-java-sdk-bundle/1.12.262/aws-java-sdk-bundle-1.12.262.jar -P /opt/spark/jars",
                          "wget https://github.com/GoogleCloudDataproc/hadoop-connectors/releases/download/v2.2.17/util-hadoop-hadoop3-2.2.17.jar -P /opt/spark/jars"])

@task(
    task_config=Spark(
        # this configuration is applied to the spark cluster
        spark_conf={
            "spark.driver.memory": "1000M",
            "spark.executor.memory": "1000M",
            "spark.executor.cores": "1",
            "spark.executor.instances": "2",
            "spark.driver.cores": "1",
        },
        executor_path="/usr/bin/python3",
        applications_path="local:///usr/local/bin/entrypoint.py",
    ),
    limits=Resources(mem="2000M"),
    cache_version="1",
    container_image=spark_image,
)
def hello_spark(partitions: int) -> float:
    print("Starting Sparkfk wifth Partitions: {}".format(partitions))
    n = 100000 * partitions
    sess = flytekit.current_context().spark_session
    count = (
        sess.sparkContext.parallelize(range(1, n + 1), partitions).map(f).reduce(add)
    )
    pi_val = 4.0 * count / n
    print("Pi val is :{}".format(pi_val))
    return pi_val


def f(_):
    x = random.random() * 2 - 1
    y = random.random() * 2 - 1
    return 1 if x**2 + y**2 <= 1 else 0


@task(cache_version="1", container_image=spark_image)
def print_every_time(value_to_print: float, date_triggered: datetime.datetime) -> int:
    print("My printed value: {} @ {}".format(value_to_print, date_triggered))
    return 1


@workflow
def wf(triggered_date: datetime.datetime = datetime.datetime.now()) -> float:
    """
    Using the workflow is still as any other workflow. As image is a property of the task, the workflow does not care
    about how the image is configured.
    """
    pi = hello_spark(partitions=50)
    print_every_time(value_to_print=pi, date_triggered=triggered_date)
    return pi
```

## Tracking Issue
https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_

### CODE REVIEW

File: flytekit/image_spec/image_spec.py [hunk 1]
```diff
@@ -120,6 +122,48 @@ def exist(self) -> bool:
     def __hash__(self):
         return hash(asdict(self).__str__())
 
+    def run_commands(self, commands: Union[str, List[str]]):
+        """
+        Run a list of commands during the building process.
+        """
+        if self.commands is None:
```
Comment (kumare3) [inline flytekit/image_spec/image_spec.py:129]: Lets not add any mutable apis to ImageSpec, ImageSpec should be immutable by default.

If you want to add new attributes, that should be like a builder.

ImageSpec.with_run_commands() -> ImageSpec

Comment (pingsutw) [reply flytekit/image_spec/image_spec.py]: done

**Review by cosmicBboy — DISMISSED** [review]

File: plugins/flytekit-envd/tests/test_image_spec.py [hunk 2]
```diff
@@ -25,6 +27,7 @@ def test_image_spec():
 
 def build():
     base(image="cr.flyte.org/flyteorg/flytekit:py3.8-latest", dev=False)
+    run(commands=["echo hello"])
```
Comment (eapolinario) [inline plugins/flytekit-envd/tests/test_image_spec.py:30]: Curious: does this work in a real image build? If we have more than one command, does it have to use a list?

Comment (pingsutw) [reply plugins/flytekit-envd/tests/test_image_spec.py:30]: The commands ["echo hello", "wget ..."] are the equivalent of
```Dockerfile
# Dockerfile
RUN echo hello
RUN wget ...
```
Yes, you can definitely use `&&` as well. like commands["echo hello && wget ..."]

File: plugins/flytekit-envd/flytekitplugins/envd/image_builder.py [hunk 3]
```diff
@@ -52,6 +52,8 @@ def create_envd_config(image_spec: ImageSpec) -> str:
     packages = [] if image_spec.packages is None else image_spec.packages
     apt_packages = [] if image_spec.apt_packages is None else image_spec.apt_packages
     env = {"PYTHONPATH": "/root", _F_IMG_ID: image_spec.image_name()}
+    commands = [] if image_spec.commands is None else image_spec.commands
```
Comment (eapolinario) [inline plugins/flytekit-envd/flytekitplugins/envd/image_builder.py:55]: Commands run before or after packages? I'm assuming that apt packages run before commands.

Comment (pingsutw) [reply plugins/flytekit-envd/flytekitplugins/envd/image_builder.py:55]: Command is always run after packages even if you change the order here. 
`apt_packages` -> `packages` -> `command`

**Comment by wild-endeavor** [comment]
does this go down the road a little bit of recreating what a dockerfile is?  won't one day people want to add additional commands and stuff?  i thought imagespec supported a full dockerfile does it not?

**Review by eapolinario — DISMISSED** [review]

**Comment by eapolinario** [comment]
> does this go down the road a little bit of recreating what a dockerfile is? won't one day people want to add additional commands and stuff? i thought imagespec supported a full dockerfile does it not?

It recreates a portion of Dockerfiles with a lot of structure. For example, you can't have multi-stage builds, you have commands run in arbitrary places.

File: flytekit/image_spec/image_spec.py [hunk 4]
```diff
@@ -120,6 +122,51 @@ def exist(self) -> bool:
     def __hash__(self):
         return hash(asdict(self).__str__())
 
+    def with_commands(self, commands: Union[str, List[str]]) -> "ImageSpec":
+        """
+        Run a list of commands during the building process.
```
Comment (kumare3) [inline flytekit/image_spec/image_spec.py:127]:
```suggestion
        Builder that returns a new imagespec with additional list of commands that will be executed during the building process.
```

File: flytekit/image_spec/image_spec.py [hunk 5]
```diff
@@ -120,6 +122,51 @@ def exist(self) -> bool:
...
+            new_image_spec.packages = []
+
+        if isinstance(packages, List):
+            new_image_spec.packages.extend(packages)
+        else:
+            new_image_spec.packages.append(packages)
+
+        return new_image_spec
+
+    def with_apt_packages(self, apt_packages: Union[str, List[str]]) -> "ImageSpec":
```
Comment (kumare3) [inline flytekit/image_spec/image_spec.py:155]: additional apt packages

File: flytekit/image_spec/image_spec.py [hunk 6]
```diff
@@ -120,6 +122,51 @@ def exist(self) -> bool:
...
+        """
+        Install python packages during the building process.
+        """
+        new_image_spec = copy.deepcopy(self)
+        if new_image_spec.apt_packages is None:
+            new_image_spec.apt_packages = []
+
+        if isinstance(apt_packages, List):
+            new_image_spec.apt_packages.extend(apt_packages)
```
Comment (kumare3) [inline flytekit/image_spec/image_spec.py:164]: why this difference?

File: flytekit/image_spec/image_spec.py [hunk 7]
```diff
@@ -120,6 +122,51 @@ def exist(self) -> bool:
...
+        if isinstance(commands, List):
+            new_image_spec.commands.extend(commands)
+        else:
+            new_image_spec.commands.append(commands)
+
+        return new_image_spec
+
+    def with_packages(self, packages: Union[str, List[str]]) -> "ImageSpec":
+        """
+        Install python packages during the building process.
```
[the remark this answers is not in the document]
Comment (pingsutw) [reply flytekit/image_spec/image_spec.py]: done, thanks

File: flytekit/image_spec/image_spec.py (hunk 4 above)
Comment (pingsutw) [reply flytekit/image_spec/image_spec.py]: updated

File: flytekit/image_spec/image_spec.py (hunk 5 above)
Comment (pingsutw) [reply flytekit/image_spec/image_spec.py:155]: updated

**Review by eapolinario — APPROVED** [review]

**Pull request merged**